### PR TITLE
fix: Load TEI in TrackerCapture app trigger update query

### PR DESCRIPTION
**Pull request by vietnguyen**
https://jira.dhis2.org/browse/DHIS2-9230

- Removed `AbstractDeviceRenderTypeMapDeserializer`
- Add function `programAttributeStore.getAttributes( programs )`  and  `entityTypeAttributeStore.getAttributes( trackedEntityTypes)`
Those new functions extract TrackedEntityAttribute  directly from hibernate, no mapping/deserialization needed. It also uses one `IN` condition instead of n queries by `programid` as in old code. 

This fix considered as a work around, it doesn't fix the root cause. Maybe the cause is because of jsonb column deserialization causing additional update queries.  Need more time to investigate and work on that.

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramTrackedEntityAttributeStore.java [hunk 1]
```diff
@@ -39,4 +42,6 @@
     extends IdentifiableObjectStore<ProgramTrackedEntityAttribute>
 {
     ProgramTrackedEntityAttribute get( Program program, TrackedEntityAttribute attribute );
+
+    List<TrackedEntityAttribute> getAttributesByProgramsWithDataRead( List<Program> programs, User user );
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramTrackedEntityAttributeStore.java:46]: 1) Please add Javadoc
2) I would rename this method to `getAttributes` and explain the ACL part in the java docs

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityTypeAttributeStore.java [hunk 2]
```diff
@@ -28,13 +28,14 @@
...
+import java.util.List;
+
+public interface TrackedEntityTypeAttributeStore
+    extends IdentifiableObjectStore<TrackedEntityTypeAttribute>
 {
-    public SectionRenderTypeDeserialize()
-    {
-        super( SectionRenderingObject::new );
-    }
+    List<TrackedEntityAttribute> getAttributesByTrackedEntityTypesWithDataRead( List<TrackedEntityType> trackedEntityTypes, User user );
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityTypeAttributeStore.java:40]: 1) Please add Javadoc
  2) I would rename this method to getAttributes and explain the ACL part in the java docs

File: dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonDeviceRenderTypeMap.java [hunk 3]
```diff
@@ -28,38 +28,32 @@
...
 public class JsonDeviceRenderTypeMap extends JsonBinaryType
 {
-    private Class<? extends JsonDeserializer> deserializer;
+    private Class<? extends RenderingObject> renderType;
 
     @Override
-    protected Object convertJsonToObject( String content )
+    protected JavaType getResultingJavaType( Class<?> returnedClass )
     {
-        try
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonDeviceRenderTypeMap.java:45]: I'm not very familiar with this renderer, but this look like a drastic change. Can you comment what has been done here?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramTrackedEntityAttributeStore.java [hunk 4]
```diff
@@ -66,4 +75,21 @@ public ProgramTrackedEntityAttribute get(Program program, TrackedEntityAttribute
             .addPredicate( root -> builder.equal( root.get( "program" ), program ) )
             .addPredicate( root -> builder.equal( root.get( "attribute" ), attribute ) ) );
     }
+
+    @Override
+    public List<TrackedEntityAttribute> getAttributesByProgramsWithDataRead( List<Program> programs, User user )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramTrackedEntityAttributeStore.java:80]: Do we have a test for this function?

File: dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonDeviceRenderTypeMap.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/hibernate/jsonb/type/JsonDeviceRenderTypeMap.java:45]: In an attempt to fix the issue, I have removed AbstractDeviceRenderTypeMapDeserializer which I have added to support XML deserialization of RenderType. So the content of the method `JsonDeviceRenderTypeMap.convertMapToObject()` is actually copied from the deserializer. Even if this doesn't solve the issue, I think it reduces the complexity of the code.
This is covered by the test `MetadataImportServiceTest.testSerializeDeviceRenderTypeMap()`

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramTrackedEntityAttributeStore.java (hunk 4 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/hibernate/HibernateProgramTrackedEntityAttributeStore.java]: Added unit test

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityTypeAttributeStore.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/trackedentity/TrackedEntityTypeAttributeStore.java]: Fixed and added unit test

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramTrackedEntityAttributeStore.java (hunk 1 above)
Comment (vietnguyen) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/ProgramTrackedEntityAttributeStore.java]: Fixed and added unit test

**Comment by vietnguyen** [comment]
> Test coverage for this PR is quite low (10%), do you think we can improve it?

@luciano-fiandesio I added unit test. Also moved the data read check out of the DAO functions to the `DefaultTrackedEntityAttributeService. getAllUserReadableTrackedEntityAttributes()` function. I think it should be faster that way.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityAttributeService.java [hunk 5]
```diff
@@ -283,13 +288,20 @@ else if ( trackedEntityAttribute.hasOptionSet() && !trackedEntityAttribute.isVal
...
+                programs.stream().filter( program -> aclService.canDataRead( user, program ) ).collect( Collectors.toList()) ) );
+        }
 
-        attributes.addAll( trackedEntityTypes.stream().filter( trackedEntityType -> aclService.canDataRead( user, trackedEntityType ) ).collect( Collectors.toList() )
-            .stream().map( TrackedEntityType::getTrackedEntityAttributes ).flatMap( Collection::stream ).collect( Collectors.toSet() ) );
+        if ( trackedEntityTypes != null && !trackedEntityTypes.isEmpty() )
+        {
+            attributes.addAll( entityTypeAttributeStore.getAttributes(
+                trackedEntityTypes.stream().filter( trackedEntityType -> aclService.canDataRead( user, trackedEntityType ) ).collect(
+                    Collectors.toList()) ));
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackedEntityAttributeService.java:303]:
```suggestion
                    Collectors.toList() ) ) );
```

**Pull request merged**